daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -1257,20 +1259,102 @@ private[lf] object Speedy {
     }
   }
 
+  /** KTryCatchHandler marks the kont-stack to allow unwinding when throw is executed. If
+    * the continuation is entered normally, the environment is restored but the handler is
+    * not executed.  When a throw is executed, the kont-stack is unwound to the nearest
+    * enclosing KTryCatchHandler (if there is one), and the code for the handler executed.
+    */
+  private[speedy] final case class KTryCatchHandler(
+      machine: Machine,
+      handler: SExpr,
+  ) extends Kont
+      with SomeArrayEquals {
+
+    val envSize = machine.env.size
+
+    private val savedBase = machine.markBase()
+    private val frame = machine.frame
+    private val actuals = machine.actuals
+
+    // we must restore when catching a throw, or for normal execution
+    def restore() = {
+      machine.restoreBase(savedBase);
+      machine.restoreFrameAndActuals(frame, actuals)
+    }
+
+    def execute(v: SValue) = {
+      restore();
+      machine.returnValue = v
+    }
+  }
+
+  /** If unwinding the kont-stack fails to find a KTryCatchHandler, we initiate evaluation
+    * of the messageFunction contained in the exception payload to produce a text
+    * message, after first pushing a KUnhandledException continuation which will
+    * actually throw the scala exception DamlEUnhandledException containing the message.
+    */
+  private[speedy] final case class KUnhandledException(
+      machine: Machine
+  ) extends Kont
+      with SomeArrayEquals {
+    def execute(payload: SValue) = {
+      payload match {
+        case SValue.SText(message) =>
+          throw DamlEUnhandledException(message)
+        case x =>
+          throw SErrorCrash(s"KUnhandledException, expected Text got $x")
+      }
+    }
+  }
+
+  private[speedy] def throwUnhandledException(machine: Machine, payload: SValue) = {
```

Comment:
What happens if the message handler throws an exception? It seems like we’ll throw the exception from the message function now. That seems a bit problematic: An unhandled exception will throw the exception from the message function instead whereas a handled exception won’t (at least not necessarily) even evaluate the message function.

that seems somewhat inconsistent. I haven’t yet checked tests and spec to see what is "correct" here. @sofiafaro-da wdyt?

Reply:
My understanding is the semantics say the message function is called only when explicitly demanded by the any-exception-message builtin. i.e. not on throw, and not on catch.

I guess the question is: Is it a good idea to call the any-exception-message for an unhandled exception?
Not sure if the semantics really answers this.

Reply:
The semantics doesn't say anything about what you do with an unhandled exception value. It only guarantees that a message function is associated with the exception type, but doesn't precompute the message (because then it would have to store it somewhere in the transaction and that would usually lead to duplication).

When reporting an unhandled exception to the user, I think the correct way to deal with this is to compute the message in a safe way -- if an error or exception occurs during message computation, report both the original error (e.g. the type) and the fact that error message computation failed.

Reply:
I marked this as a TODO.

If we do embed the original value/type of the LF-exception in the scala exception `DamlEUnhandledException` then perhaps there is no reason to call the message function at all in this place.